theseus/optimizer/nonlinear/trust_region.py

```diff
@@ -0,0 +1,146 @@
+            max_iterations=max_iterations,
+            step_size=step_size,
+            **kwargs,
+        )
+        self._trust_region: torch.Tensor = None
+
+    def reset(
+        self,
+        trust_region_init: float = 0.5,
+        **kwargs,
+    ) -> None:
+        super().reset(**kwargs)
+        self._trust_region = trust_region_init * torch.ones(
+            self.objective.batch_size,
+            1,
+            device=self.objective.device,
+            dtype=self.objective.dtype,
+        )
+
+    # Return the computed delta
+    @abc.abstractmethod
+    def _compute_delta_impl(self) -> torch.Tensor:
+        pass
+
+    def compute_delta(self, **kwargs) -> torch.Tensor:
+        return self._compute_delta_impl()
+
+    @staticmethod
+    def _squared_norm(tensor: torch.Tensor, keepdim: bool = True) -> torch.Tensor:
+        return (tensor**2).sum(dim=1, keepdim=keepdim)
+
+    # Computes m_k(delta) as defined in Eq. (4.2) of the above reference (p. 68)
+    def _predicted_error(
+        self, previous_error: torch.Tensor, delta: torch.Tensor
+    ) -> torch.Tensor:
+        linearization = self.linear_solver.linearization
+
+        # Note that B = |A^t@ A| so that p^tBp := delta^t A^t A delta = ||Adelta||^2
+        Adelta = linearization.Av(delta)
+        grad = -linearization.Atb.squeeze(2)
+        delta_dot_grad = (delta * grad).sum(dim=1)
+
+        return (
+            previous_error
+            + delta_dot_grad
+            + 0.5 * TrustRegion._squared_norm(Adelta, keepdim=False)
+        )
+
+    def _complete_step(
+        self,
+        delta: torch.Tensor,
+        new_err: torch.Tensor,
+        previous_err: torch.Tensor,
+        accept_threshold: float = 0.0,
+        shrink_threshold: float = 0.25,
+        expand_threshold: float = 0.75,
+        shrink_ratio: float = 0.25,
+        expand_ratio: float = 2.0,
+        min_trust_region: float = 1.0e-5,
+        max_trust_region: float = 1.0e5,
+        **kwargs,
+    ) -> Optional[torch.Tensor]:
+        # "err" tensors passed as input refer to the squared norm of the
+        # error vector, as returned by self._error_metric()
+        good_params = (0.0 < shrink_ratio <= 1.0) and (expand_ratio >= 1.0)
+        good_params &= (shrink_threshold < expand_threshold) and (
+            accept_threshold < shrink_threshold
+        )
+        if not good_params:
+            raise ValueError(
+                "Invalid parameters for TrustRegionMethod. "
+                "Values must satisfy <accept/shrink>_threshold < expand_threshold, "
+                "shrink_ratio in (0, 1], and expand_ratio > 1.0."
+            )
+        pred_err = self._predicted_error(previous_err, delta)
+        rho = ((previous_err - new_err) / (previous_err - pred_err)).view(-1, 1)
+        shrink_idx = rho < shrink_threshold
+        self._trust_region = torch.where(
+            shrink_idx, self._trust_region * shrink_ratio, self._trust_region
+        )
```

Comment:
Should we add a stopping condition here if `self._trust_region` is too small?

Reply:
Maybe we need to add another vector `stopped` to indicate whether an optimization problem is still active or not.

Reply:
I think our current `converged_indices` is functionally equivalent to `stopped`.